### PR TITLE
fix: Match RoleBindings bound to ClusterRoles

### LINKED ISSUE #31: Possible gap where rolebindings are used with clusterroles
I was looking at a cluster today and noticed that there were some results I was expecting which didn't show up.  I *think* that what might be happening is that `kubectl-who-can` isn't picking up cases where a `RoleBinding` is bound to a `ClusterRole` (the other option is I'm doing something wrong here, which is quite possible :) ). 

I did the following test with a vanilla `kind` cluster.

- Create a ClusterRole which can get secrets

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  # "namespace" omitted since ClusterRoles are not namespaced
  name: secret-reader
rules:
- apiGroups: [""]
  resources: ["secrets"]
  verbs: ["get", "watch", "list"]
```

- Create a new namespace

```
kubectl create namespace test
```

- Create a new service account in the namespace

```
kubectl -n test create serviceaccount test1
```

- Create a RoleBinding in that namespace

```
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: read-secrets
  namespace: test
subjects:
- kind: ServiceAccount
  name: test1 
roleRef:
  kind: ClusterRole 
  name: secret-reader 
  apiGroup: rbac.authorization.k8s.io
```

At this point if I use `kubectl auth can-i` it shows that the service account can get secrets in that namespace.

```
kubectl -n test auth can-i get secrets --as=system:serviceaccount:test:test1
yes
```

However if I use `kubectl-who-can` like this the test account isn't in the list of subjects
```
./kubectl-who-can -n test get secrets
No subjects found with permissions to get secrets assigned through RoleBindings

CLUSTERROLEBINDING                                    SUBJECT                             TYPE            SA-NAMESPACE
cluster-admin                                         system:masters                      Group           
system:controller:clusterrole-aggregation-controller  clusterrole-aggregation-controller  ServiceAccount  kube-system
system:controller:expand-controller                   expand-controller                   ServiceAccount  kube-system
system:controller:generic-garbage-collector           generic-garbage-collector           ServiceAccount  kube-system
system:controller:horizontal-pod-autoscaler           horizontal-pod-autoscaler           ServiceAccount  kube-system
system:controller:namespace-controller                namespace-controller                ServiceAccount  kube-system
system:controller:persistent-volume-binder            persistent-volume-binder            ServiceAccount  kube-system
system:kube-controller-manager                        system:kube-controller-manager      User  
```

### PR DESCRIPTION
**Pull request by danielpacak**
This one is supposed to resolve #31 . I took also the opportunity to refactor and add more unit tests to the functionality of mapping [Cluster]RoleBindings to [Cluster]Roles.
At the end I've added the scenario to our integration test suite which reproduced the actual issue reported by @raesene

I suggest review this PR commit by commit.

### CODE REVIEW

File: pkg/cmd/list.go [hunk 1]
```diff
@@ -395,6 +397,7 @@ func (w *whoCan) GetClusterRolesFor(action Action) (clusterRoles, error) {
 
 // GetRoleBindings returns the RoleBindings that refer to the given set of Role names or ClusterRole names.
 func (w *whoCan) GetRoleBindings(roleNames roles, clusterRoleNames clusterRoles) (roleBindings []rbac.RoleBinding, err error) {
+	// TODO I'm wondering if GetRoleBindings should be invoked at all when the --all-namespaces flag is specified?
```
Comment (lizrice) [inline pkg/cmd/list.go]: Interesting! I think this comes down to the semantics of `--all-namespaces`, which I now think is ambiguous: 

* If it means "who can verb resource in **any** namespace" then yes, I think we should be looking at all the possible rolebindings. 
* If it means "who can verb resource in **all** namespace" then we *could* limit to just clusterRoles, but then we'd be missing the possibility of someone who has this ability, not through a clusterRole but by being bound to a Role for every namespace

The first of these options seems more useful to me anyway. Either way, we should clarify it in the usage.

File: test/integration_test.go [hunk 2]
```diff
@@ -74,69 +90,151 @@ func TestIntegration(t *testing.T) {
...
 
-	_, err = clientRBAC.ClusterRoles().Create(&rbac.ClusterRole{
-		ObjectMeta: meta.ObjectMeta{Name: "get-logs"},
+	_, err = clientRBAC.Roles(core.NamespaceDefault).Create(&rbac.Role{
+		ObjectMeta: meta.ObjectMeta{Name: "scale-workloads"},
 		Rules: []rbac.PolicyRule{
 			{
-				Verbs:           []string{"get"},
-				NonResourceURLs: []string{"/logs"},
+				APIGroups: []string{""},
```
Comment (lizrice) [inline test/integration_test.go]: If we wanted to support CRDs, would we need to allow the user to specify the APIGroup?

File: pkg/cmd/list.go [hunk 3]
```diff
@@ -49,6 +49,11 @@ NONRESOURCEURL is a partial URL that starts with "/".`
 
   # List who can access the URL /logs/
   kubectl who-can get /logs`
+
+	// RoleKind is the RoleRef's Kind referencing a Role.
+	RoleKind = "Role"
```
Comment (lizrice) [inline pkg/cmd/list.go]: I'm OK with this as it's unlikely to change, but is there not a definition we could pull in from kubernetes somewhere?

File: pkg/cmd/list.go [hunk 4]
```diff
@@ -344,167 +356,96 @@ func (w *whoCan) printAPIAccessWarnings(warnings []string) {
...
 	if err != nil {
 		return
 	}
 
-	for _, roleBinding := range rbl.Items {
-		if w.r.match(&roleBinding.RoleRef) {
-			glog.V(1).Info(fmt.Sprintf("Match found: roleRef: %v", roleBinding.RoleRef))
+	for _, roleBinding := range list.Items {
+		if _, ok := clusterRoleNames[roleBinding.RoleRef.Name]; ok {
+			//if w.clusterRoleBindingMatches(&roleBinding, clusterRoleNames) {
```
Comment (lizrice) [inline pkg/cmd/list.go]: I doubt you meant to leave this comment here

File: test/integration_test.go (hunk 2 above)
Comment (danielpacak) [reply test/integration_test.go]: That's a good point. We do not currently resolve APIGroup:

```
$ kubectl auth can-i get deployments.extensions
yes

$ kubectl who-can get deployments.extensions
Error: resolving resource: the server doesn't have a resource type "deployments.extensions"
```

File: pkg/cmd/list.go (hunk 3 above)
Comment (danielpacak) [reply pkg/cmd/list.go]: Unfortunately not. I did search in project's dependencies and it seems that it's common to repeatedly type in `"Role"` and `"CusterRole"` literals all over the place ;)

File: pkg/cmd/list.go (hunk 1 above)
Comment (danielpacak) [reply pkg/cmd/list.go]: So maybe we should have a dedicated flags `--any-namespace` and `--all-namespaces` to make it less ambiguous? Plus what you mentioned a better usage docs.

**Comment by danielpacak** [comment]
Hey @lizrice thanks for the review. Let's merge this bulky PR. I have created dedicated issues #38 #39 to follow up on you comments.

**Pull request merged**